overlord/ifacestate/ifacemgr.go

```diff
@@ -73,54 +73,116 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, extra []interfa
 	return m, nil
 }
 
+func initialConnectAttributes(s *state.State, plugSnap string, plugName string, slotSnap string, slotName string) (map[string]interface{}, error) {
+	// Combine attributes from plug and slot, store them in connect task.
+	// They will serve as initial attributes for the prepare- hooks.
+	var snapst snapstate.SnapState
+	var err error
+	attrs := make(map[string]interface{})
+
+	if err = snapstate.Get(s, plugSnap, &snapst); err != nil {
+		return nil, err
+	}
+
+	snapInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+	if plug, ok := snapInfo.Plugs[plugName]; ok {
+		for k, v := range plug.Attrs {
+			attrs[k] = v
+		}
+	} else {
+		return nil, fmt.Errorf("Snap %q has no plug named %q", plugSnap, plugName)
+	}
+
+	if err = snapstate.Get(s, slotSnap, &snapst); err != nil {
+		return nil, err
+	}
+	snapInfo, err = snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+	if slot, ok := snapInfo.Slots[slotName]; ok {
+		for k, v := range slot.Attrs {
+			attrs[k] = v
+		}
+	} else {
+		return nil, fmt.Errorf("Snap %q has no slot named %q", slotSnap, slotName)
+	}
+
+	return attrs, err
+}
+
 // Connect returns a set of tasks for connecting an interface.
 //
 func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*state.TaskSet, error) {
 	// TODO: Store the intent-to-connect in the state so that we automatically
 	// try to reconnect on reboot (reconnection can fail or can connect with
 	// different parameters so we cannot store the actual connection details).
+
+	// Create a series of tasks:
+	//  - prepare-plug-<plug> hook
+	//  - prepare-slot-<slot> hook
+	//  - connect task
+	//  - confirm-slot-<slot> hook
+	//  - confirm-plug-<plug> hook
+	// The tasks run in sequence (are serialized by WaitFor).
+	// The prepare- hooks collect attributes set via snapctl iset. The attributes set by prepare-plug
+	// hook can be read by both prepare-plug and prepare-slot hooks. All the attributes collected by
+	// first two hooks are available to the connect task and both confirm- hooks for reading.
+	summary := fmt.Sprintf(i18n.G("Connect %s:%s to %s:%s"),
+		plugSnap, plugName, slotSnap, slotName)
+	connectInterface := s.NewTask("connect", summary)
+	initialContext := map[string]interface{}{"connect-task": connectInterface.ID()}
+
 	plugHookSetup := &hookstate.HookSetup{
 		Snap:     plugSnap,
 		Hook:     "prepare-plug-" + plugName,
 		Optional: true,
 	}
-	summary := fmt.Sprintf(i18n.G("Prepare connection of plug %s:%s"), plugSnap, plugName)
-	preparePlugConnection := hookstate.HookTask(s, summary, plugHookSetup, nil)
+	summary = fmt.Sprintf(i18n.G("Prepare connection of plug %s:%s, hook %s"), plugSnap, plugName, plugHookSetup.Hook)
```

Comment:
This is a bit long and redundant (_"Prepare ... plug foo:bar, hook prepare-plug-bar"_). If the hook name is necessary and makes the experience more obvious, it's probably better to have only the hook name:

`"Run hook %s of snap %q"`

Same below.